### PR TITLE
Wait longer time when finding some segments are in recovery in NotifyingCommitPrepared()

### PR DESCRIPTION
**Pull request by interma**
**TLDR:**
The PR introduced a new GUC `dtx_phase2_recovery_retry_second`, it used as the timeout when segments are in recovery statue. (the default value is 15m, usually bigger than `dtx_phase2_retry_second`)

And no need to doc the new GUC since no related content about `dtx_phase2_retry_second` (they are advanced GUCs for superuser)

(Below is the previous detailed discussion)
### background
When some segments are failure, `NotifyingCommitPrepared()` does several retries until `dtx_phase2_retry_second` (default is 60s) reaches, and reports a PANIC finally:
https://github.com/greenplum-db/gpdb/blob/00d0fef6ee6d0d293810fed373226c038be6947e/src/backend/cdb/cdbtm.c#L639-L642

But if the segments are in recovery state (usually, this process takes a long time), reporting PANIC prematurely may "scare" user. (Note: this PANIC is no harm, just make cluster to restart to redo it again)
So, waiting longer time is better in the scenario (then user's queries hang instead of PANIC).

### enhancement
In our previous discussion, I plan to do a more precise timeout control by tracing the progress of recovery state (like we did in `checkIfFailedDueToNormalRestart()` in ftsprobe.c: checking the recovery state).
But unlike the FTS logic (only need check one segment), here we need a complex structure to trace all segments' xlog progress, e.g.
```
dbid1 => 1359/BDA3AB0
dbid2 => 1420/FDA3AB0
...
dbidn => XX/XX
```
And the basic idea is to keep waiting if any seg's progress is in advance of a small time window (compared to last state). 

But in the coding phase, l found it makes the related code more complex:
* mixed several retry logic for recovery:
    * in any failure, we retry dispatch and it calls create_gang()
    * create_gang() retries until `gp_gang_creation_retry_count` reaches when any segment is in recovery
    * and we want to introduce a new retry logic in `checkIfFailedDueToNormalRestart()`
    * mixing them together isn't easy to understand
* need to expose/transfer the original libpq message (record which segment is in recovery and what's current wal position) in all dispatch's functions, make code ugly
* hard to write a regress test
* and the new behavior may still confuse user: query may hang a very long time and no progress report (in psql)

Based all, I chose a simpler timeout mechanism in this PR:
Just introduce another timeout value (in my PR, it's double of `dtx_phase2_retry_second`, or a new GUC is also acceptable?) for recovery.
It can decrease the possibility of PANIC and easy to understand the behavior, especially we also have enhancement in the gprecovery utility later (check the segment recovery progress).

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [x] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [x] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/cdb/cdbtm.c [hunk 1]
```diff
@@ -626,13 +628,25 @@ doNotifyingCommitPrepared(void)
 			 * restore the previous value, which is reset to 0 in errfinish.
 			 */
 			MemoryContextSwitchTo(oldcontext);
+
+			/* double timedout_second when some segs are in recovery/reset state */
+			ErrorData  *errdata = CopyErrorData();
+			if (strstr(errdata->detail, _(SEG_RECOVERY_SUFFIX_STRING)))
+			{
+				timeout_second = dtx_phase2_retry_second*2;
```
Comment (interma) [inline src/backend/cdb/cdbtm.c:636]: For reviewer: just double the timeout value here, may introduce a new GUC is better.

**Comment by interma** [comment]
And I will add test later.

**Comment by interma** [comment]
@ashwinstar please help to review my "enhancement" section when you are free, thanks!

File: src/backend/cdb/cdbtm.c (hunk 1 above)
Comment (ashwinstar) [reply src/backend/cdb/cdbtm.c:636]: I don't think just doubling the timeout does any better. If that was the case we would have just asked customers to double the `dtx_phase2_retry_second` and problem would be gone which is not the case. Seems fine to introduce another GUC with much higher timeout for recovery case like 15 mins or so as default.

**Comment by ashwinstar** [comment]
> @ashwinstar please help to review my "enhancement" section when you are free, thanks!

Looks fine to me. Please add test for the added functionality (any new functionality must have test associated with it).

File: src/backend/cdb/cdbtm.c (hunk 1 above)
Comment (interma) [reply src/backend/cdb/cdbtm.c:636]: Thank you, Ashwin.
I am going to change it.

**Comment by ashwinstar** [comment]
@SunilS26 Please help to review this PR

**Comment by interma** [comment]
@ashwinstar @SunilS26 thank you!
I am working on adding test case and making my pipeline to green now.
No hurry to review it. (anytime when you are free)

**Comment by interma** [comment]
Ready for review @SunilS26 
Thanks!

File: src/backend/cdb/cdbtm.c [hunk 2]
```diff
@@ -626,13 +640,27 @@ doNotifyingCommitPrepared(void)
 			 * restore the previous value, which is reset to 0 in errfinish.
 			 */
 			MemoryContextSwitchTo(oldcontext);
+
+			ErrorData  *errdata = CopyErrorData();
+			if (errdata->detail != NULL && strstr(errdata->detail, _(SEG_RECOVERY_SUFFIX_STRING)))
```
Comment (SunilS26) [inline src/backend/cdb/cdbtm.c:645]: i think "_" is added by miss in the above strstr argument "_(SEG_RECOVERY_SUFFIX_STRING)". If so please remove this.

File: src/backend/cdb/cdbtm.c [hunk 3]
```diff
@@ -626,13 +640,27 @@ doNotifyingCommitPrepared(void)
...
+			{
+				/*
+				 * let's wait longer time when segments are in recovery/reset
+				 */
+				if (timeout_second < dtx_phase2_recovery_retry_second)
+					timeout_second = dtx_phase2_recovery_retry_second;
+				ereport(WARNING,
+						(errmsg("Segments are in recovery/reset in the distributed transaction 'Commit Prepared' broadcast"),
```
Comment (SunilS26) [inline src/backend/cdb/cdbtm.c:653]: Should we even log the current timeout_second value along with errmsg log ? . 
Just a thought , because when we are going through the log's for some similar issues we'll be knowing the exact timeout at that time, which helps in debugging.

**Review by SunilS26 — COMMENTED** [review]
Just started understanding the changes and issues. Some minor comments....

**Comment by interma** [comment]
> Just started understanding the changes and issues. Some minor comments....

Thank you @SunilS26, no hurry to review this PR, just as your pace.
And you can get all backgrounds in our internal Jira: MPP-32378.

File: src/backend/utils/misc/guc_gp.c [hunk 4]
```diff
@@ -4142,6 +4143,18 @@ struct config_int ConfigureNamesInt_gp[] =
 		NULL, NULL, NULL
 	},
 
+	{
+		{"dtx_phase2_recovery_retry_second", PGC_SUSET, GP_ARRAY_TUNING,
+			gettext_noop("Maximum time for which coordinator tries to finish a prepared transaction when segments are in recovery/reset state"),
+			gettext_noop("The timer is for recovery state, so it's bigger than dtx_phase2_retry_second."
+						 " Coordinator keeps retrying the finish-prepared operation"
+						 " until this timeout (seconds)."),
+			GUC_SUPERUSER_ONLY |  GUC_NO_SHOW_ALL | GUC_NOT_IN_SAMPLE | GUC_UNIT_S
```
Comment (SunilS26) [inline src/backend/utils/misc/guc_gp.c:4152]: It's better to  expose GUC to user.We should remove "GUC_NO_SHOW_ALL".

File: src/backend/cdb/cdbtm.c (hunk 2 above)
Comment (SunilS26) [reply src/backend/cdb/cdbtm.c:645]: Same logic we used in fts as well. closing this query.

**Review by SunilS26 — COMMENTED** [review]
Me and Ashwin had a discussion on the current design. What we felt was that even creating a new GUC and increasing the timeout to 15min, will still be not helpful. Rather let's remove the introduction of new GUC and bump the existing GUC value to some larger timeout of 8hr or  10hr, and try infinitely till that time.

**Comment by interma** [comment]
> Me and Ashwin had a discussion on the current design. What we felt was that even creating a new GUC and increasing the timeout to 15min, will still be not helpful. Rather let's remove the introduction of new GUC and bump the existing GUC value to some larger timeout of 8hr or 10hr, and try infinitely till that time.

Thanks Sunil (and Ashwin).
Yes, as you said, a unified timeout may not suit all scenarios.

In our previous discussion (in the internal Jira), the wise solution is: choose a flexible timeout time (based on the wal gap between primary and mirror), this logic is placed into the `gprecovery` side (may it's wip now).

But I think a good behavior of the PR is: tell the latest status for user (not hang here silently), so user has the choice to know what happened:
```
WARNING:  the distributed transaction 'Commit Prepared' broadcast failed to one or more segments. Retrying ... try 1
DETAIL:  gid=16451, state=Retry Commit Prepared
NOTICE:  Releasing segworker group to retry broadcast.

// <= we can refine the message: may tell it will panic later
WARNING:  Segments are in recovery/reset in the distributed transaction 'Commit Prepared' broadcast
HINT:  It keeps retrying until successfully or reaching dtx_phase2_recovery_retry_second. 

WARNING:  the distributed transaction 'Commit Prepared' broadcast failed to one or more segments. Retrying ... try 2
DETAIL:  gid=16451, state=Retry Commit Prepared
...
```

Glad to hear your further ideas, thanks!

**Comment by SunilS26** [comment]
> > Me and Ashwin had a discussion on the current design. What we felt was that even creating a new GUC and increasing the timeout to 15min, will still be not helpful. Rather let's remove the introduction of new GUC and bump the existing GUC value to some larger timeout of 8hr or 10hr, and try infinitely till that time.
> 
> Thanks Sunil (and Ashwin). Yes, as you said, a unified timeout may not suit all scenarios.
> 
> In our previous discussion (in the internal Jira), the wise solution is: choose a flexible timeout time (based on the wal gap between primary and mirror), this logic is placed into the `gprecovery` side (may it's wip now).
> 
> And another good behavior of the PR is: tell the latest status for user (not hang here silently), so user has the choice to know what happened:
> 
> ```
> WARNING:  the distributed transaction 'Commit Prepared' broadcast failed to one or more segments. Retrying ... try 1
> DETAIL:  gid=16451, state=Retry Commit Prepared
> NOTICE:  Releasing segworker group to retry broadcast.
> 
> // <= we can refine the message: may tell it will panic later
> WARNING:  Segments are in recovery/reset in the distributed transaction 'Commit Prepared' broadcast
> HINT:  It keeps retrying until successfully or reaching dtx_phase2_recovery_retry_second. 
> 
> WARNING:  the distributed transaction 'Commit Prepared' broadcast failed to one or more segments. Retrying ... try 2
> DETAIL:  gid=16451, state=Retry Commit Prepared
> ...
> ```
> 
> Based all, I think I can change the behavior of PR to:
> 
> * Doesn't introduce the new Guc
> * When found segments are in recovery state, use a longer timeout (but I think hours to too long...) and show latest status message (let user know the panic will happen later)
> * No recovery state, still use current short timeout
> 
> I hope to hear your further ideas, thanks!

Logging warning looks good to me but we need to think about the frequency as well. As too many retry attempts may flood the log file.

**Comment by ashwinstar** [comment]
> Yes, as you said, a unified timeout may not suit all scenarios.

I think the main summary of discussion with Sunil was no need to distinguish between n/w outage vs primary restart scenario. In both the cases Coordinator should retry the commit/abort and it can't take any other action anyways. Even PANIC and restart the coordinator the situation doesn't improve and it has to restart the same operation. Hence, seems better to modify the current GUC itself to have larger timeout and introduce some periodic notifications to user for these retries. (Only if we had the capability to act similar to FTS of detecting recovery making progress and such based on LSN value we would have done things more intelligently - else its fine to just have larger retries and not cause more issues with coordinator PANIC which doesn't help the situation).

**Comment by interma** [comment]
Ashwin said:
> Even PANIC and restart the coordinator the situation doesn't improve and it has to restart the same operation. 

I rechecked the related code: But I think PANIC is a good behavior.
Here are my thoughts (please correct me if have errors):
* Option1 - Current logic: retry until reaching `dtx_phase2_retry_second` (default: 60s), then PANIC to trigger QD's postmaster restart.
    * After postmaster restart, it dives into a different code path (recovery process): `DtxRecoveryMain()` -> `recoverTM()` -> `recoverInDoubtTransactions()` -> `doNotifyCommittedInDoubt()` -> dispatch `DTX_PROTOCOL_COMMAND_RECOVERY_COMMIT_PREPARED` to QEs again
    * This code path doesn't goto `doNotifyingCommitPrepared()`, just report FATAL rather than PANIC when failure: https://github.com/greenplum-db/gpdb/blob/033c4f3c63566d3639d8c12c0b4ad0d023d939c0/src/backend/cdb/cdbdtxrecovery.c#L86 and recovery bgworker will restart and redo it again (equal to retry forever until successfully)
    * The system doesn't accept new query request until the recovery is done
* Option2 - Retry longer time: no PANIC just retries again and again in QD.
    * It's ok for the retry-transaction, but other queries may be blocked by it (tuple lock or intent lock).
    * More importantly, user may consider the system runs normally and are confused as to why many queries hang silent.  

Based on all, I think we just need to set a bigger default value for `dtx_phase2_retry_second`, e.g. 10min.
So:
* If recovery is done quickly, user no need to experience a PANIC
* If recovery needs a long time, PANIC to cause system to down into the recovery process. It's as expected and system goes to normal after recovery is done.

Please help to review, Ashwin and Sunil.
Thanks!

**Comment by ashwinstar** [comment]
> Based on all, I think we just need to set a bigger default value for dtx_phase2_retry_second, e.g. 10min.

Seems fair and fine approach. We can always iterate on it based on user input in future. So, yes lets increase the timeout to 15 mins and rest the issue for now. Thanks for all the inputs and thinking on the same.

**Comment by interma** [comment]
Thanks Ashwin, I think @SunilS26 is working on reproducing the original issue and thinking about it.
If he has no other ideas, I will revert most of the PR and just change default value of `dtx_phase2_retry_second ` to 15min.

**Comment by SunilS26** [comment]
> Thanks Ashwin, I think @SunilS26 is working on reproducing the original issue and thinking about it. If he has no other ideas, I will revert most of the PR and just change default value of `dtx_phase2_retry_second ` to 15min.
Yes, please go ahead with the changes. The approach looks good to me as well.

**Comment by interma** [comment]
Thank you Ashwin and Sunil, I will make the changes soon.

**Comment by interma** [comment]
@SunilS26 I have reverted my previous changes and enlarge `dtx_phase2_retry_second` to 600s (with detailed comment for it).
Please help to review again, thanks.

File: src/backend/utils/misc/guc_gp.c [hunk 5]
```diff
@@ -4129,6 +4129,18 @@ struct config_int ConfigureNamesInt_gp[] =
...
+	 * (Details are in doNotifyingCommitPrepared()).
+	 * Then the postmaster will restart and goes into recovery process.
+	 *
+	 * So, a small value may make user confused: why my postmaster restarts; but a big value
+	 * is also not good: the txn keeps retrying in dispatch, it may block other txns.
+	 *
+	 * After a long discussion: https://github.com/greenplum-db/gpdb/pull/15632, we choose a
+	 * compromise default value: 600s(10min) here.
+	 */
 	{
```
Comment (SunilS26) [inline src/backend/utils/misc/guc_gp.c:4144]: +1 nice explanation

**Review by SunilS26 — APPROVED** [review]
LGTM :-) 
Thanks for all the exploration and explanation!

**Comment by interma** [comment]
Thanks All, have merged it.
And I am going to port it to 6x soon.
